src/promdens/promdens.py

```diff
@@ -163,6 +163,48 @@ def calc_field_envelope(self, t: np.ndarray) -> np.ndarray:
             return field
 
 
+    def pulse_wigner(self, tprime, de):
+        """
+        Wigner transform of the pulse. The current implementation uses the pulse envelope formulation to simplify calculations.
+        The integral is calculated numerically. Analytic formulas could be implemented here.
+
+        :param tprime: the excitation time t' (a.u.)
+        :param de: excitation energy (a.u.)
+        :return: Wigner pulse transform
+        """
+        # setting an adaptive integration step according to the frequency of the integrand oscillations (de - omega)
+        loc_omega = self.omega + 2*self.lchirp*tprime
+        # If de == loc_omega, we're in resonance and there are no oscillations,
+        # integrate only the envelope intensity so we can set a larger integration step
+        if de == loc_omega:
+            T = 2*np.pi/loc_omega
+        else:
+            T = 2*np.pi/(np.min([np.abs(de - loc_omega), loc_omega]))
```

Comment:
It might be nice to comment on this formula, the use of np.min here seems non-obvious.

Reply:
It is non-obvious. I use it because if I keep only `de - loc_omega`, then the numerical integral is extremely demanding when we are far from resonance. This is often the case when you have multiple states that are well separated. I will need to think about it a bit and come up with something better. Let's keep it like that and I will address at least part of it in a different pull request.